### PR TITLE
Add files via upload

### PR DESCRIPTION
**Pull request by szeng7**

### CODE REVIEW

**Comment by szeng7** [comment]
@safipatel for the experiments.ipynb file where you have Expotentiated Gradient constraint and Adversarial Learning, can you make sure it works on a basic classifier (logreg, random forest, anything works) with the newer dataset I put in?

In train.ipynb, there are a few cells at the beginning where I'm pulling in the train data like such

```
data_dict = pd.read_pickle('cleaned_data_stacked_and_dict.pkl')
train_df = pd.concat([data_dict['2006'], data_dict['2008'], data_dict['2010']], ignore_index=True)
X_train = train_df.drop(columns=['mortality_ten_years', 'mortality_five_years', 'year_death', 'HHID', 'PN'])  # Adjust target column
y_train = train_df['mortality_ten_years']
X_train = X_train.fillna(-999)
```

I'm working on some intermediary data points of differing values of WD and then once I have those all prepped out, we can easily run a bunch of experiments with those to see how distributional shift affects the fairness metrics (and most importantly, its behavior-- you can see what I just added to the README in the next steps where I explain it a bit more) and then how these fairness methods can potentially mitigate that

**Comment by szeng7** [comment]
main thing with this adjusted dataset is that I merge some of the years together, so now any given row contains

person, current year properties, properties from 2 years ago, properties from 4 years ago. So the model should be able to perform a bit better using the change in history (maybe a neural net would recognize this better? I think a current linear classifier can't do much with it unless we explicitly calculate it out)

**Comment by safipatel** [comment]
Okay so first, I added a requirements.txt so if you want to just use the commands in the install script or just run pip install directly, we can be on the same page for that.

I've also added a lot of ground work that can be extended to more trials. I moved the training ground work to a function which is useable for different experiments (unfortunately the FairML expotentiatedgradient doesn't allow for warm_start or multiple fit() calls, so we don't have a graph of accuracy over iterations). More importantly, I also added a `display_metrics()` function which can be used throughout.

Another important thing to note: imputation and normalization performed a lot better. I apply it with the Pipeline. We see the accuracy of 0.75ish capable with imputation/normalization but more importantly, it is far more stable if you look at the graphs. I think I can improve this even further if I (1) label all the fields that are categorical (2) make them ordinal features (because currently race or other features are 1 2 5 9 as their numbers so the model might get confused numerically by that) (3) for the categorical features use most_frequent imputation rather than mean

More comments about exponentiated gradient: I had to drop any rows where the sensitive attribute was NAN because it didn't allow that. And as discussed above, I couldn't use my generic training function since it doesn't allow calling fit() multiple times. But overall, it seems to perform pretty well I think even with the constraint. I don't see any difference in demographic (statistical) parity for ex. but I believe that's because it's only on the training data. I can do it on the testing data too but I just have to fix the dropping behavior for the test dataframe.

I'm going to do the AdversarialTraining one next but I had to install torch so it'll be a little bit.

**Comment by safipatel** [comment]
Firstly, I changed handling of categorical vs numerical data. I went through all the columns and noted the ones that are actual floats, and said the rest are just categories. Then later when I'm doing the imputation and normalization, I can (1) make sure that its encoded ordinally so there's no weird numeric gaps between the categories, then I can also use most frequent imputation for categories and mean imputation for floats. Doing this, the MLP got to 0.95ish on the training. I've also displayed more of the metrics run across train, test, test ood. And finally, the adversarial looks a bit better, I've changed the model and also probably from the categorical vs numeric handling.
